formula: trap CMake FetchContent usage instead of using FETCHCONTENT_FULLY_DISCONNECTED

alebcay opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Follow-up to #17075. With this change, when a formula attempts to use `FetchContent_MakeAvailable`, the following error is presented:

```
-- Configuring built-in curl...
CMake Error at /opt/homebrew/Library/Homebrew/trap_fetchcontent_provider.cmake:4 (message):
  Refusing to populate dependency 'zlib' with FetchContent while building in
  Homebrew, please add the dependency as a dependency or resource in the
  formula
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.29.3/share/cmake/Modules/FetchContent.cmake:2017:EVAL:1 (trap_fetchcontent_provider)
  /opt/homebrew/Cellar/cmake/3.29.3/share/cmake/Modules/FetchContent.cmake:2017 (cmake_language)
  cmake/zlib_external.cmake:15 (FetchContent_MakeAvailable)
  CMakeLists.txt:214 (include)


-- Configuring incomplete, errors occurred!
```

(This is using a modified version of the `cpr` formula to try the scenario where `FetchContent_MakeAvailable` gets used; normally this formula will rely on brewed dependencies.)

Formulae can opt out by adding flag `-DCMAKE_PROJECT_TOP_LEVEL_INCLUDES=` (essentially unsetting the variable). Maybe we can provide something less unwieldy/better named to accomplish this.

Seeking feedback especially about:
- The error message and what end users or formula authors can do about it
- Name and location of the `.cmake` file - for now I just dropped it in `HOMEBREW_LIBRARY_PATH`, maybe it makes sense to be in some subdirectory, or in the Cellar somewhere (include this file in the `cmake` formula/bottle rather than as part of `brew`)

MikeMcQuaid commented:
> Formulae can opt out by adding flag `-DCMAKE_PROJECT_TOP_LEVEL_INCLUDES=` (essentially unsetting the variable). Maybe we can provide something less unwieldy/better named to accomplish this.

Yeh, ideally there'd be a dedicated CMake variable for this `-DHOMEBREW_ALLOW_FETCHCONTENT` or something?

> * The error message and what end users or formula authors can do about it

I think it's great!

MikeMcQuaid commented on Library/Homebrew/trap_fetchcontent_provider.cmake:
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.24) # Dependency providers introduced in CMake 3.24
```
Maybe just put in a `cmake` subdirectory?

MikeMcQuaid commented on Library/Homebrew/trap_fetchcontent_provider.cmake:
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.24) # Dependency providers introduced in CMake 3.24
+
+macro(trap_fetchcontent_provider method depName)
```
Could gate this on a `HOMEBREW_ALLOW_FETCHCONTENT` variable or similar?

carlocab commented on Library/Homebrew/trap_fetchcontent_provider.cmake:
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.24) # Dependency providers introduced in CMake 3.24
+
+macro(trap_fetchcontent_provider method depName)
+    message(FATAL_ERROR "Refusing to populate dependency '${depName}' with FetchContent while building in Homebrew, please add the dependency as a dependency or resource in the formula")
```
```suggestion
    message(FATAL_ERROR "Refusing to populate dependency '${depName}' with FetchContent while building in Homebrew, please use a formula dependency or add a  resource to the formula")
```

alebcay commented:
- Moved `.cmake` file into `cmake` subdirectory
- Added `HOMEBREW_ALLOW_FETCHCONTENT` option, which can be set to `ON` to bypass this
- Adjusted message text

MikeMcQuaid reviewed (approved):
Yeh, this looks perfect, thanks again @alebcay!

carlocab commented on Library/Homebrew/cmake/trap_fetchcontent_provider.cmake:
```diff
@@ -0,0 +1,14 @@
+option(HOMEBREW_ALLOW_FETCHCONTENT "Allow FetchContent to be used in Homebrew builds" OFF)
+
+macro(trap_fetchcontent_provider method depName)
+    if (NOT HOMEBREW_ALLOW_FETCHCONTENT)
```
I wonder if we can/should just return early if this is set.

alebcay replied:
I'm open to this and it is possible, but I don't think it's as idiomatic in CMake (nor does it offer as much of a readability benefit as it does in Ruby):

```cmake
if (HOMEBREW_ALLOW_FETCHCONTENT)
    return()
endif()

macro(trap_fetchcontent_provider method depName)
    [...]
endif()
```

alebcay commented:
Adjusted the scope of the conditional - we probably should just avoid creating the dependency provider altogether if we want to bypass.

carlocab commented on Library/Homebrew/cmake/trap_fetchcontent_provider.cmake:
```diff
@@ -0,0 +1,14 @@
+
+option(HOMEBREW_ALLOW_FETCHCONTENT "Allow FetchContent to be used in Homebrew builds" OFF)
+
+if (NOT HOMEBREW_ALLOW_FETCHCONTENT)
+    macro(trap_fetchcontent_provider method depName)
+        message(FATAL_ERROR "Refusing to populate dependency '${depName}' with FetchContent while building in Homebrew, please use a formula dependency or add a resource to the formula.")
+    endmacro()
+
+    cmake_language(
+        SET_DEPENDENCY_PROVIDER trap_fetchcontent_provider
+        SUPPORTED_METHODS FETCHCONTENT_MAKEAVAILABLE_SERIAL
+    )
+endif()
```
```suggestion
if (HOMEBREW_ALLOW_FETCHCONTENT)
    return()
endif()

macro(trap_fetchcontent_provider method depName)
    message(FATAL_ERROR "Refusing to populate dependency '${depName}' with FetchContent while building in Homebrew, please use a formula dependency or add a resource to the formula.")
endmacro()

cmake_language(
    SET_DEPENDENCY_PROVIDER trap_fetchcontent_provider
    SUPPORTED_METHODS FETCHCONTENT_MAKEAVAILABLE_SERIAL
)
```
Doesn't work?

alebcay replied:
It does work, see my earlier comment. I don't see early returns often in CMake, but sure we can do it; I'm not particularly opinionated on it one way or another..

carlocab replied:
Ah, sorry; didn't see it since the conversation was marked as resolved. To me reducing indentation always makes things easier to read, but I'm not going to push you on it.

But, re idiomatic CMake, I see lots of early `return`s in the built-in CMake modules (e.g. see scripts in `$(brew --prefix cmake)/share/cmake/Modules`).

MikeMcQuaid reviewed (approved):
Looks good, thanks again @alebcay!
